File: love/game/room.lua
-- game/room.lua
-- Room state: tracks the 4 card slots and per-turn rules.
--
-- Each turn, 4 cards are dealt into the room. The player must take 3 of them
-- (one at a time), leaving the 4th to carry over into the next room.
--
-- Slots can be sparse (e.g., {nil, Card, nil, Card}) after cards are resolved,
-- so all iteration uses `for i = 1, 4` instead of `ipairs`.
--
-- Constraint: only one health potion may be used per room.

local Class = require("lib.class")

local Room = Class:extend()

function Room:init()
    self.slots = { nil, nil, nil, nil } -- exactly 4 slots, nil = empty
    self.resolved_count = 0             -- how many cards the player has taken this room
    self.potion_used = false            -- has a potion been used this turn?
    self.initial_count = 0              -- how many cards were dealt (for completion check)
end

-- Set up the room with cards at specific slot indices.
-- `slot_cards` is a table keyed by slot index (1-4), e.g., { [2] = card, [3] = card }.
function Room:set_cards(slot_cards)
    self.slots = { nil, nil, nil, nil }
    local count = 0
    for i = 1, 4 do
        if slot_cards[i] then
            self.slots[i] = slot_cards[i]
            count = count + 1
        end
    end
    self.resolved_count = 0
    self.potion_used = false
    self.initial_count = count
end

-- Mark a card as resolved (taken by the player). Returns the card.
function Room:resolve_card(index)
    local card = self.slots[index]
    self.slots[index] = nil
    self.resolved_count = self.resolved_count + 1
    return card
end

-- Is this room complete?
-- Standard room (4 cards): take 3, leave 1 carry-over.
-- Short room (< 4 cards, near end of dungeon): take all but 1, or take all if only 1.
function Room:is_complete()
    if self.initial_count >= 4 then
        return self.resolved_count >= 3
    elseif self.initial_count > 1 then
        return self.resolved_count >= self.initial_count - 1
    else
        return self.resolved_count >= 1
    end
end

-- Get the carry-over card and its slot index.
-- Returns card, slot_index (or nil, nil if empty).
function Room:get_remaining_card()
    for i = 1, 4 do
        if self.slots[i] then
            return self.slots[i], i
        end
    end
    return nil, nil
end

-- Get all current cards (non-nil slots). Used when avoiding a room.
function Room:get_all_cards()
    local cards = {}
    for i = 1, 4 do
        if self.slots[i] then
            table.insert(cards, self.slots[i])
        end
    end
    return cards
end

-- How many cards are still in the room?
function Room:card_count()
    local count = 0
    for i = 1, 4 do
        if self.slots[i] then count = count + 1 end
    end
    return count
end

-- Is a specific slot occupied?
function Room:has_card_at(index)
    return self.slots[index] ~= nil
end

-- Clear the room entirely.
function Room:clear()
    self.slots = { nil, nil, nil, nil }
    self.resolved_count = 0
    self.potion_used = false
    self.initial_count = 0
end

return Room

File: ebitengine/game/room.go
package game

import "github.com/thesimpledev/radianthorizon/entities"

// Room tracks the 4 card slots and per-turn rules.
type Room struct {
	Slots         [4]*entities.Card // nil = empty
	ResolvedCount int
	PotionUsed    bool
	InitialCount  int
}

// NewRoom creates an empty room.
func NewRoom() *Room {
	return &Room{}
}

// SetCards sets up the room with cards at specific slot indices.
// slotCards is keyed by index 0-3.
func (r *Room) SetCards(slotCards [4]*entities.Card) {
	r.Slots = slotCards
	r.ResolvedCount = 0
	r.PotionUsed = false
	count := 0
	for i := 0; i < 4; i++ {
		if r.Slots[i] != nil {
			count++
		}
	}
	r.InitialCount = count
}

// ResolveCard marks a card as taken. Returns the card.
func (r *Room) ResolveCard(index int) *entities.Card {
	card := r.Slots[index]
	r.Slots[index] = nil
	r.ResolvedCount++
	return card
}

// IsComplete checks if this room is finished.
func (r *Room) IsComplete() bool {
	if r.InitialCount >= 4 {
		return r.ResolvedCount >= 3
	} else if r.InitialCount > 1 {
		return r.ResolvedCount >= r.InitialCount-1
	}
	return r.ResolvedCount >= 1
}

// GetRemainingCard returns the carry-over card and its slot index.
// Returns nil, -1 if empty.
func (r *Room) GetRemainingCard() (*entities.Card, int) {
	for i := 0; i < 4; i++ {
		if r.Slots[i] != nil {
			return r.Slots[i], i
		}
	}
	return nil, -1
}

// GetAllCards returns all non-nil cards currently in the room.
func (r *Room) GetAllCards() []*entities.Card {
	var cards []*entities.Card
	for i := 0; i < 4; i++ {
		if r.Slots[i] != nil {
			cards = append(cards, r.Slots[i])
		}
	}
	return cards
}

// CardCount returns how many cards are in the room.
func (r *Room) CardCount() int {
	count := 0
	for i := 0; i < 4; i++ {
		if r.Slots[i] != nil {
			count++
		}
	}
	return count
}

// HasCardAt checks if a slot is occupied.
func (r *Room) HasCardAt(index int) bool {
	return index >= 0 && index < 4 && r.Slots[index] != nil
}

// Clear empties the room.
func (r *Room) Clear() {
	r.Slots = [4]*entities.Card{}
	r.ResolvedCount = 0
	r.PotionUsed = false
	r.InitialCount = 0
}
